File: services/front/src/types/messages.ts
import type { Player } from "../types/room"
import type { GameMap, GameStat, PlayerData } from "../types/map"

export type PlayersMessage = {
    at: "players";
    players: Player[];
};

export type AuthMessage = {
  at: "auth";
  token: string;
}

export type ReadyMessage = {
  at: "ready";
}

export type ChatMessage = {
  at: "chat";
  user_id: number;
  message: string;
  username: string;
  timestamp: string;
};

export type UpdateMessage = {
  at: 'update';
  map: GameMap;
  turn: number;
  stat: [PlayerData, GameStat];
}

File: services/rooms/src/app_types/messages.py
from typing import Literal, TypedDict

from app_types.in_messages import (
    AuthMessage,
    ColorMessage,
    MoveMessage,
    ReadyMessage,
)
from app_types.out_messages import AuthConfirmMessage, PlayersMessage, StartMessage, UpdateMessage


class ChatMessage(TypedDict):
    at: Literal["chat"]
    user_id: int
    message: str
    username: str
    timestamp: str


InMessage = AuthMessage | ReadyMessage | MoveMessage | ColorMessage | ChatMessage


OutMessage = PlayersMessage | AuthConfirmMessage | StartMessage | UpdateMessage | ChatMessage
